server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java

```diff
@@ -0,0 +1,243 @@
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.model.mailbox;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.BinaryOperator;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.james.mailbox.model.MailboxACL;
+import org.apache.james.mailbox.model.MailboxACL.EntryKey;
+import org.apache.james.mailbox.model.MailboxACL.Rfc4314Rights;
+import org.apache.james.util.OptionalUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Multimap;
+
+public class Rights {
+    public enum Right {
+        Administer(MailboxACL.Right.Administer),
+        Expunge(MailboxACL.Right.PerformExpunge),
+        Insert(MailboxACL.Right.Insert),
+        Lookup(MailboxACL.Right.Lookup),
+        Read(MailboxACL.Right.Read),
+        Seen(MailboxACL.Right.WriteSeenFlag),
+        T_Delete(MailboxACL.Right.DeleteMessages),
+        Write(MailboxACL.Right.Write);
+
+        private final MailboxACL.Right right;
+
+        Right(MailboxACL.Right right) {
+            this.right = right;
+        }
+
+        @JsonValue
+        public char asCharacter() {
+            return right.asCharacter();
+        }
+
+        public MailboxACL.Right getRight() {
+            return right;
+        }
+
+        public static Optional<Right> forRight(MailboxACL.Right right) {
+            return OptionalUtils.ifEmpty(
```

Comment:
there's a semantic problem with this ifEmpty : you don't expect a "ifSomething execute my lambda" to actually return the optional

Reply:
What name would you propose?

Reply:
I didn't said the name is wrong. It should nonetheless return void.
So you would have two separated instructions (3 steps) : 
1. transformation
2. ifEmpty + log
3. return

Reply:
This method was first intended to write one liners more easily (originaly used for logging empty optionals with Cassandra related code). Returning the original optional value makes it possible.

I propose **peekIfEmpty** as a name.

Reply:
it makes no sense to me : what do you peek then ? Maybe the problem is about trying to be too generic. Given we already have peek, orElse, orElseThrow, what problems ifEmpty method solves except logging ?
If we agree it's only useful for logging, we could have a special case for that : logIfEmpty that takes a lambda.

Reply:
In one hand, I agree @mbaechler ``ifEmpty`` should be a void but in the other hand it's so conveniant it just return the value.
``peekIfEmpty`` would be definitly worst. 
What about ``identityWithSideEffectIfEmpty`` :trollface:

Reply:
or simply executeIfEmpty?